lib/phoenix_live_view/diff.ex

```diff
@@ -567,15 +619,110 @@ defmodule Phoenix.LiveView.Diff do
-          traverse(rendered, {nil, %{}}, pending, components, template, false)
+  defp traverse_keyed(
+         entries,
+         previous_prints,
+         pending,
+         components,
+         template,
+         path,
+         changed?,
+         stream?
+       ) do
+    diff = %{}
+    new_prints = %{}
+
+    # TODO: we could optimize the diff further and not send an empty @keyed when the
+    # map_size(previous_prints) == map_size(new_prints)
+    {diff, count, new_prints, pending, components, template} =
+      Enum.reduce(entries, {diff, 0, new_prints, pending, components, template}, fn
+        # it's an existing entry
+        %KeyedComprehensionEntry{fingerprint: {fingerprint, new_vars}, render: render},
+        {diff, index, new_prints, pending, components, template}
+        when is_map_key(previous_prints, fingerprint) and not stream? ->
+          %{vars: previous_vars, index: previous_index, child_prints: child_prints} =
+            Map.fetch!(previous_prints, fingerprint)
+
+          vars_changed =
+            Enum.reduce(new_vars, Map.put(previous_vars, :__changed__, %{}), fn
+              {key, value}, acc ->
+                Phoenix.Component.assign(acc, key, value)
+            end)
+            |> Map.fetch!(:__changed__)
+
+          {child_diff, child_prints, pending, components, template} =
+            traverse(
+              render.(vars_changed, changed?),
+              child_prints,
+              pending,
+              components,
+              template,
+              [fingerprint | path],
+              changed?
+            )
+
+          new_prints =
+            Map.put(new_prints, fingerprint, %{
+              index: index,
+              vars: new_vars,
+              child_prints: child_prints
+            })
+
+          if child_diff == %{} or child_diff == nil do
+            # the entry did not change, we can skip it
+            if previous_index != index do
+              # the entry moved, annotate it with the previous index
+              {Map.put(diff, index, previous_index), index + 1, new_prints, pending, components,
+               template}
+            else
+              {diff, index + 1, new_prints, pending, components, template}
+            end
+          else
+            child_diff =
+              if previous_index != index do
+                [previous_index, child_diff]
+              else
+                child_diff
+              end
+
+            {Map.put(diff, index, child_diff), index + 1, new_prints, pending, components,
+             template}
+          end
 
-        {diff, {pending, components, template}}
+        # it's a new entry
+        %KeyedComprehensionEntry{fingerprint: {fingerprint, vars}, render: render},
+        {diff, index, new_prints, pending, components, template} ->
+          {child_diff, child_prints, pending, components, template} =
+            traverse(
+              render.(%{}, false),
+              %{},
```

Comment:
```suggestion
              {nil, %{}},
```

Perhaps store it in an env var...

However my suggestion would be to get any fingerprint from anywhere in this comprehension and use it, to allow memory sharing (if we are lucky)

Reply:
Or instead compute it through the comprehension.